eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/ZmqWatcher.scala

```diff
@@ -239,13 +251,134 @@ class ZmqWatcher(chainHash: ByteVector32, blockCount: AtomicLong, client: Extend
+        adjustAnchorOutputChange(unsignedTx, commitTx, fundTxResponse.amountIn + Transactions.AnchorOutputsCommitmentFormat.anchorAmount, currentFeerate, targetFeerate, dustLimit)
+      }).flatMap(claimAnchorTx => {
+        val claimAnchorSig = keyManager.sign(claimAnchorTx, localFundingPubKey, Transactions.TxOwner.Local, commitmentFormat)
+        val signedClaimAnchorTx = Transactions.addSigs(claimAnchorTx, claimAnchorSig)
+        val commitInfo = ExtendedBitcoinClient.PreviousTx(signedClaimAnchorTx.input, signedClaimAnchorTx.tx.txIn.head.witness)
+        client.signTransaction(signedClaimAnchorTx.tx, Seq(commitInfo))(singleThreadExecutionContext)
+      }).flatMap(signTxResponse => {
+        client.publishTransaction(signTxResponse.tx)(singleThreadExecutionContext)
+      })
+    }
+  }
+
+  def publishHtlcTx(currentFeerate: FeeratePerKw, targetFeerate: FeeratePerKw, dustLimit: Satoshi, signingKit: TransactionSigningKit.HtlcTxSigningKit): Future[ByteVector32] = {
+    import signingKit._
+    if (targetFeerate <= currentFeerate) {
+      val localSig = keyManager.sign(txWithInput, localHtlcBasepoint, localPerCommitmentPoint, Transactions.TxOwner.Local, commitmentFormat)
+      val signedHtlcTx = addHtlcTxSigs(txWithInput, localSig, signingKit)
+      log.info("publishing htlc tx without adding inputs: txid={}", signedHtlcTx.tx.txid)
+      client.publishTransaction(signedHtlcTx.tx)(singleThreadExecutionContext)
+    } else {
+      log.info("publishing htlc tx with additional inputs: commit input={}:{} target feerate={}", txWithInput.input.outPoint.txid, txWithInput.input.outPoint.index, targetFeerate)
+      val txNotFunded = txWithInput.tx.copy(txIn = Nil, txOut = txWithInput.tx.txOut.head.copy(amount = dustLimit) :: Nil)
+      val htlcTxWeight = signingKit match {
+        case _: TransactionSigningKit.HtlcSuccessSigningKit => commitmentFormat.htlcSuccessWeight
+        case _: TransactionSigningKit.HtlcTimeoutSigningKit => commitmentFormat.htlcTimeoutWeight
+      }
+      // NB: bitcoind will add at least one P2WPKH input.
+      val weightRatio = htlcTxWeight.toDouble / (txNotFunded.weight() + Transactions.claimP2WPKHOutputWeight)
+      client.fundTransaction(txNotFunded, FundTransactionOptions(targetFeerate * weightRatio, lockUtxos = true, changePosition = Some(1)))(singleThreadExecutionContext).map(fundTxResponse => {
+        log.info(s"added ${fundTxResponse.tx.txIn.length} wallet input(s) and ${fundTxResponse.tx.txOut.length - 1} wallet output(s) to htlc tx spending commit input=${txWithInput.input.outPoint.txid}:${txWithInput.input.outPoint.index}")
+        // We add the HTLC input (from the commit tx) and restore the HTLC output.
+        val txWithHtlcInput = fundTxResponse.tx.copy(
+          txIn = txWithInput.tx.txIn ++ fundTxResponse.tx.txIn,
+          txOut = txWithInput.tx.txOut ++ fundTxResponse.tx.txOut.tail
+        )
```

Comment:
Why don't you merge the outputs here? That would simplify `adjustHtlcTxChange` I think.

Reply:
We can't merge the outputs, this is an HTLC-tx.
Our peer has signed it with `SIGHASH_SINGLE | SIGHASH_ANYONECANPAY` to let us add additional inputs and outputs, but we can't modify the first input nor the first output otherwise we invalidate their signature...
I'll add a comment to explain that.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1702/commits/378108ab76f2e936b3984c298403d5dc2aba519f